### ISSUE TITLE
PHP randomly fails to open other required files

### ISSUE THREAD

**Issue by strarsis** [docker/for-win#2005]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have submitted Diagnostics
  - [x] I have included the Diagnostics ID in this Issue
  - Windows Version: Windows 10 Pro 1803 17134.1
  - Docker for Windows Version: Version 18.03.1-ce-win65 (17513)

### Expected behavior
No issues with PHP loading other required files.

### Actual behavior
With each odd or so request, PHP fatal errors with PHP failing to open a required PHP file.

Example:
````
PHP Fatal error:  require(): Failed opening required '/srv/www/web/wp/wp-includes/script-loader.php' (include_path='.:/usr/local/lib/php') in /srv/www/web/wp/wp-settings.php on line 192
````

### Information

Docker Diagnostics ID: DB6939B5-5158-41B4-8EA0-9DC8A82F5590/2018-05-04_22-32-13

### Steps to reproduce the behavior
  1. Start a Docker container from library php image (7.x fpm) with directories mounted from WSL (Bash on Windows) / native NTFS drive.
  2. (Ideally a WordPress sample installation) Request some pages, PHP should fail opening some of the required files quite regularly and randomly.

How is the disk/directory shared in Hyper-V - CIFS? May PHP have a buffer/async read related issue?

**Comment by jbroutier** [comment]
Same issue here. 

- Docker 18.03.1-ce-win65 (17513)
- Windows 10 1803 build 17134.48

PHP require() randomly fails, especially when executing multiple requests in parallel.
Maybe a file lock problem ?

**Comment by strarsis** [comment]
@jbroutier: Yes! Probably the reason.
In order to start `mariadb` I have to use this config:
````
innodb_flush_method=littlesync
innodb_use_native_aio=OFF
log_bin=ON
````
So `littlesync` or `nosync` have to be used and AIO disabled.
Maybe there are equivalent options in PHP?

**Comment by c-lan** [comment]
@jasonbivins I don't believe this is related to compose, the issue also affects `docker run` with volume mounts. Also, Windows 1803 is currently available publicly, not only for Windows Insider members. I presume large number of users could be affected.

What I observed, randomly one or more files mounted through volumes becomes inaccessible. Windows SMB server returns STATUS_OBJECT_NAME_NOT_FOUND for that file and keeps doing so until the session is terminated and new one established. The file in question is only being read, no writes or deletes are attempted. In the meantime, another SMB session (e.g. opened through Explorer) works fine and the file is accessible there. Restarting SMB server or remounting the CIFS mount solves the problem intermittently. Disabling more advanced SMB server features (e.g. oplocks) or changing mount flags does not seem to affect anything.

**Comment by andreas-ateles** [comment]
Having similar random file not found issues in shared volumes (#969, #1856).

* Docker 18.03.1-ce-win65 (17513)
* Host: Windows 10 Pro 64-bit, version 1803, build 17134.1
* Applications failing in Linux containers: PHP in Apache 2.4, yarn, WebPack, etc.

Copying the files to a non-shared volume and running from there works without issues. But that's not a viable solution for a development workflow.

**Comment by andreas-ateles** [comment]
@jasonbivins this is not a compose related issue. It occurs even if the container is started with the regular docker command.

Windows 10 version 1803 is not an insider build any more: 
> public release on April 30, and began to roll out on May 8

So probably should remove the `area/windows insider` tag?

This bug makes docker for windows completely unusable when accessing host files. Maybe it should have some kind of priority?

**Comment by strarsis** [comment]
@andreas-ateles: Yes, this bug makes it also practically unusable for WordPress sites.
Saving a new post (page) results in HTTP 500 with the content lost.
It is usable for developping a theme using browsersync but here it ends already.

**Comment by jbroutier** [comment]
You can also tell PHP-FPM to use only one process by setting the `pm` directive to `static` and `pm.maxchildren` to 1 in the pool configuration file, and thus prevent any concurrent access to the files.

An example PHP-FPM pool configuration:

```
[www]
user = www-data
group = www-data

listen = 127.0.0.1:9000

pm = static
pm.max_children = 1
```

The path of the default pool config file in the official PHP-FPM image is `/usr/local/etc/php-fpm.d/www.conf`. Simply use a Dockerfile with the following content and rebuild the image:

```
COPY ./www.conf /usr/local/etc/php-fpm.d/
````

Of course it's a bit slower than the default config, but it works.

**Comment by strarsis** [comment]
@jbroutier: Yes! This extra config file fixes the issue. It is a bit slower but I don't need to reload multiple times to get the full page loaded.

**Comment by cjamieso** [comment]
I'm running into similar issues with other php images (apache rather than fpm).  Seeing random failures when disk access is required: require/include/file_get_contents or when the class autoloader runs.  Running composer (typically always on an autoload) from the command line also seems to trigger the random failures as well.

The errors only occur in 1803 (1709 works perfectly), so it must have to do with an undocumented Hyper-V change or some of the SMB related changes in 1803.

Has anyone found a more general solution for all php containers?  Will use the rsync method above in the interim.  As long as the files aren't part of a mounted volume, everything seems to work fine.

**Comment by elebeaup** [comment]
I have a similar issues with node image when building application in shared volume and more specifically when compiling sass (node-sass plugin). I have a file not found error but the file exist and is accessible.

If I build application in no shared volumes, there works without issues.
If I use SMB mounting with vers=2.0 instead of vers=3.0, I have no issues. This may impact performance but at least everything seems works...

**Comment by strarsis** [comment]
@elebeaup: Are you using Docker for Windows or Docker Toolbox? If so, how have you switched SMB from `vers` `3.x` to `2.0`? I would try to downgrade this, too for getting around this issue without resorting to special (PHP) config.

**Comment by elebeaup** [comment]
I use Docker for Windows and I have switched smb version (from vers 3.0 to 2.0).

**Comment by joel1st** [comment]
@elebeaup - I can confirm that I have also been seeing the same things with node builds (must be a filesystem bug related to multiple threads trying to access the same file).

**Comment by plastbox** [comment]
Steps to reproduce: [https://github.com/docker/for-win/issues/969#issuecomment-411429451](url)

**Comment by wadewomersley** [comment]
Same here, requests randomly fail with a file not found error in PHP.

**Comment by strarsis** [comment]
@plastbox: The link seems to be broken.

**Comment by sazo** [comment]
Thanks to @Werelds and https://github.com/docker/for-win/issues/969#issuecomment-411429451. Helped me a lot. I just want to share my solution when using docker-compose. 

php-fpm uses the "code-volume" created below. I have added a vendor-helper to speed things a bit up only downside is you need to run the vendor-helper after each composer install. 

docker-compose.override.yml
```
services:
  php-fpm:
      volumes:
          - code-volume:/var/www/html
          - vendor-data:/var/www/html/vendor

  vendor-helper:
      image: busybox
      command: '/bin/sh -c "rm -rf /data/* && cp -r /vendor-data/* /data"'
      volumes:
        - vendor-data:/data
        - ./vendor:/vendor-data

volumes:
  vendor-data:
    driver: local
  code-volume:
      external: true
```

Create the CIFS v2.0 code-volume used in the docker-compose.override.yml
```
[subfolder] = C/Users/userfolder/path/to/project
[username] = user
[password] = password
[domain] = DESKTOP-6BOGBNF

docker volume create --driver local --opt type=cifs --opt device="//10.0.75.1/[subfolder]" --opt o="rw,vers=2.0,sec=ntlmsspi,cache=strict,username=[username],pass=[password],domain=[domain],uid=0,noforceuid,gid=0,noforcegid,addr=10.0.75.1,file_mode=0777,dir_mode=0777,iocharset=utf8,nounix,serverino,mapposix,nobrl,mfsymlinks,noperm,rsize=1048576,wsize=1048576,echo_interval=60,actimeo=1" code-volume
```

**Comment by elebeaup** [comment]
Like what @blurayne says in https://github.com/docker/for-win/issues/2285#issuecomment-408377309, it would be nice to add a feature to place default mount options somewhere in the Docker for Windows config (or to fix the bug with CIFS v2.1 and up)

**Comment by aftabnaveed** [comment]
In my case I was using php swoole which was spawning more than one worker at the same time reading the same set of files in parallel. Changing that to 1 for development fixed my issue. Looks like multiple processes accessing the same file even for reading is causing this issue.

**Comment by andreas-ateles** [comment]
https://github.com/docker/for-win/issues/969#issuecomment-420285666 shows how to configure CIFS v2.0 from docker-compose:

```
services:
  web:
    build: .
    command: bundle exec rails s -p 3000 -b '0.0.0.0'
    volumes:
      - code:/myapp
    ports:
      - "3000:3000"

volumes:
  code:
    driver: local
    driver_opts:
      type: cifs
      device: '//10.0.75.1/C/path/to/code'
      o: 'rw,relatime,vers=2.0,sec=ntlmsspi,cache=strict,username=brian,password=***,domain=***,uid=0,noforceuid,gid=0,noforcegid,addr=10.0.75.1,file_mode=0755,dir_mode=0777,iocharset=utf8,nounix,mapposix,nobrl,mfsymlinks,noperm,rsize=1048576,wsize=1048576,echo_interval=60,actimeo=1'
```

**Comment by nicholasio** [comment]
Having same issue with docker edge (2.0.0-win77). Changing php-fpm pool config to static and one process solved the issue but slows down things quite a bit.

**Comment by gtsiou** [comment]
You can also try to use https://github.com/EugenMayer/docker-sync/wiki/docker-sync-on-Windows since that did the trick for me. For vagrant you can just switch to NFS with a couple of plugins.

The issue is caused by the way files are synced between the systems (and vagrant has the same issue) with the latest SMB.

**Comment by nicholasio** [comment]
@gtsiou I just switched back to vagrant and I'm not having any issues related to this.

**Comment by gtsiou** [comment]
> @gtsiou I just switched back to vagrant and I'm not having any issues related to this.

@nicholasio Interesting! My vagrant setup for replicating this is latest Vagrant on windows 10 with VirtualBox and SMB v3. Are you using something similar? If this does not happen for you would be a good indication that something else is to blame.

**Comment by nicholasio** [comment]
> > @gtsiou I just switched back to vagrant and I'm not having any issues related to this.
> 
> @nicholasio Interesting! My vagrant setup for replicating this is latest Vagrant on windows 10 with VirtualBox and SMB v3. Are you using something similar? If this does not happen for you would be a good indication that something else is to blame.

I'm just using latest Vagrant with Latest Virtual Box. I'm using with VVV - https://github.com/varying-vagrant-vagrants/vvv/

**Comment by mabdelfattah** [comment]
> @gtsiou You can also try to use https://github.com/EugenMayer/docker-sync/wiki/docker-sync-on-Windows since that did the trick for me. For vagrant you can just switch to NFS with a couple of plugins.
> 
> The issue is caused by the way files are synced between the systems (and vagrant has the same issue) with the latest SMB.

Unfortunately, the developer decided to stop his WSL support!
https://github.com/EugenMayer/docker-sync/issues/609#issuecomment-443681666


@andreas-ateles Do you mind sharing more details about how to install/configure CIFS on WSL?

**Comment by crissi** [comment]
I added opcache to my docker  and I stopped getting this error. Have not doubled check if this was the real reason, but give it a try. Remember validate_timestamps 1 to make it dev friendly, https://www.reddit.com/r/PHP/comments/62kne2/comment/dfoqsuw



Få Outlook til iOS<https://aka.ms/o0ukef>

________________________________
Fra: Mahmoud Abdel-Fattah <notifications@github.com>
Sendt: onsdag, december 19, 2018 6:56 PM
Til: docker/for-win
Cc: Christian; Manual
Emne: Re: [docker/for-win] PHP randomly fails to open other required files (#2005)


@gtsiou<https://github.com/gtsiou> You can also try to use https://github.com/EugenMayer/docker-sync/wiki/docker-sync-on-Windows since that did the trick for me. For vagrant you can just switch to NFS with a couple of plugins.

The issue is caused by the way files are synced between the systems (and vagrant has the same issue) with the latest SMB.

Unfortunately, the developer decided to stop his WSL support!
EugenMayer/docker-sync#609 (comment)<https://github.com/EugenMayer/docker-sync/issues/609#issuecomment-443681666>

@andreas-ateles<https://github.com/andreas-ateles> Do you mind sharing more details about how to install/configure CIFS on WSL?

**Comment by stwime** [comment]
Had this problem for a while, tried every fix to no avail. 
Fixed it today by moving everything to a new partition, setting that to shared in the docker settings and suddenly everything worked as expected. 
I guess this problem happens if mounting volumes from the C:/ drive, at least that was the case for me. 
Hope this helps!

**Comment by DirkWolthuis** [comment]
Same problem. No fix yet.

**Comment by DirkWolthuis** [comment]
Factory reset of Docker seemed to helped, it has been a couple of hours without the problem.

**Comment by DirkWolthuis** [comment]
> 
> 
> Had this problem for a while, tried every fix to no avail.
> Fixed it today by moving everything to a new partition, setting that to shared in the docker settings and suddenly everything worked as expected.
> I guess this problem happens if mounting volumes from the C:/ drive, at least that was the case for me.
> Hope this helps!

I tried this, I've got one error that a file could not be read, but besides that I've had no troubles. Will update if problems return.

**Comment by mendrik** [comment]
> PS. Anyone using WSL, maybe that is the cause of this problem?

I use WSL and got the same problem.

**Comment by mabdelfattah** [comment]
Yes, I use WSL too, and I think the problem is with Docker for Windows and WSL together

**Comment by DirkWolthuis** [comment]
Update: I've gotten a couple of more file not found errors when moved to a new partion, but nothing to bad and it's workable.

**Comment by unfalln** [comment]
I just use powershell to invoke docker rather than WSL (although I have WSL installed using the original ubuntu dist) and any volume that I specify that shares files between the Windows host and the *nix guest has the "file not found" error whenever more than one process within the guest attempts to read a file at the same time.

**Comment by DirkWolthuis** [comment]
> 
> 
> I just use powershell to invoke docker rather than WSL (although I have WSL installed using the original ubuntu dist) and any volume that I specify that shares files between the Windows host and the *nix guest has the "file not found" error whenever more than one process within the guest attempts to read a file at the same time. I can workaround this by specifying the CIFS version to <=2.0 (rather than the default 3.0.2) in the volume mount parameters, but this normally means I have to specify my windows host login username and password in the docker-compose file which is not ideal for committing to source control.

Do you have a example what to put into the compose file? Or a link where I can read more about CIFS?

**Comment by andreas-ateles** [comment]
@unfalln seems like the same problem as everyone else here. FYI: you can usually put your specific docker compose changes in the `docker-compose.override.yml` file (which could be git ignored). This is still not ideal but should work for making the CIFS version changes.

**Comment by mendrik** [comment]
actually I get the same problem starting docker on powershell. got a bunch of ajax requests hitting php scrupts that require_once the same config file and they fail randomly, with maybe 1-2 scripts executing correctly here and then.

**Comment by DirkWolthuis** [comment]
> 
> 
> actually I get the same problem starting docker on powershell. got a bunch of ajax requests hitting php scripts that require_once the same config file and they fail randomly, with maybe 1-2 scripts executing correctly here and then.

Same here, Powershell made no difference vs WSL

**Comment by DirkWolthuis** [comment]
Okay so after moving my Docker files to a other partition I have had a stable working environment for over a week. For anyone facing the troubles described inside this thread, I can recommend creating a extra partition.

**Comment by archon810** [comment]
Same problem with Version 2.0.0.3 (31259), Channel: stable, Build: 8858db3 on Windows 10 Pro Version 1809 stable (not insider).

The Wordpress dev site running on OpenSUSE + apache + PHP 7.3 is full of these random file errors that constantly break the site. 

How has Docker not prioritized this issue to critical yet?

**Comment by 22116** [comment]
Ah guys, I suggest you give up and switch on linux ¯\_(ツ)_/¯

**Comment by KristianI** [comment]
Maybe this is of interest:
https://blog.docker.com/2019/07/docker-wsl2-tech-preview/

**Comment by archon810** [comment]
/remove-lifecycle stale

Though, it's unclear if the changes in 2.2 will resolve this. Anyone know?

**Comment by andreas-ateles** [comment]
Seems like that in Docker desktop 2.2 they have replaced Samba shares with something else (gRPC FUSE). That's great! And thus it should avoid this specific problem with Samba shares that this bug is about. But if there are any similar issues on in 2.2 I don't know. Have not tried it yet.

**Comment by stephen-turner** [comment]
This should be fixed with the new filesharing implementation in 2.2.0.3, so closing this ticket. Please open a new ticket if the problem persists.

**Comment by strarsis** [comment]
@stephen-turner: Does this also apply to Docker Desktop?
Latest Docker Desktop Beta (Edge) version is `2.1.7.0 (41536)`, 
which is lower than the `2.2.0.3` version you stated.

**Comment by stephen-turner** [comment]
@strarsis There will be a new Edge version including all the fixes tomorrow.

**Comment by strarsis** [comment]
@stephen-turner: That's great, I will update when it becomes available.

**Comment by jbroutier** [comment]
@stephen-turner I just got the 2.2.1.0 (42746) update, everything is working well.
SF 5 app initialization time also dropped below 1 second 👍

**Comment by strarsis** [comment]
The issue seems to be fixed, but now I got new nightmares with cached I/O that doesn't update. 🙀

**Comment by thexpand** [comment]
How to get the new update for Docker for Windows - it says it's up to date and I cannot find any way to update it.

**Comment by strarsis** [comment]
@thexpand: You have to download/install a so called "Edge" release:
https://docs.docker.com/docker-for-windows/edge-release-notes/

**Comment by thexpand** [comment]
@strarsis Will I have to remove the previous release or I just have to install this "Edge" release on top of the other one?

**Comment by strarsis** [comment]
@thexpand: I prefer to uninstall the old one first, then reboot the system and then install the new one.
You can try and start the Edge setup, maybe it offers an upgrade.
In either case you should consider that the Docker VM is reset and you will have to re-build and re-pull the images anew (which shouldn't be an issue).
